containers/src/Data/IntMap/Internal.hs

```diff
@@ -246,288 +443,410 @@ module Data.IntMap.Internal (
+xor :: Key -> Bound t -> Word
+xor a (Bound b) = Data.Bits.xor (i2w a) (i2w b)
+
+-- | Xor the minimum and maximum keys of a tree. The most significant bit of the result indicates
+-- which bit to split on for navigation and is useful in merging maps to tell whether nodes from
+-- different maps branch at the same time.
+{-# INLINE xorBounds #-}
+xorBounds :: Bound L -> Bound R -> Word
+xorBounds (Bound min) (Bound max) = Data.Bits.xor (i2w min) (i2w max)
+
+{-# INLINE boundsDisjoint #-}
+boundsDisjoint :: Bound L -> Bound R -> Bool
+boundsDisjoint (Bound min) (Bound max) = min > max
+
+-- These are uninhabited to ensure that they are only used as type parameters.
+-- | A type tag denoting 'Node's found on left branches or minimum 'Bound's.
+data L
+-- | A type tag denoting 'Node's found on right branches or maximum 'Bound's.
+data R
+
+#if USE_TYPE_FAMILIES
+-- TODO: If we are relying on GHC features anyway, L and R could be a new kind.
+
+-- | A 'Key' stored somewhere within an 'IntMap' with a tag representing its role within that map.
+-- Following the left-to-right ascending order of keys in the map, a @'Bound' 'L'@ serves the
+-- role of a minimum (and so must be less than all other keys in the subtree where it was found),
+-- and a @'Bound' 'R'@ serves the role of maximum (and so must be greater than all other keys in
+-- the subtree where it was found).
+--
+-- Because the tag represents the relationship between a 'Bound' and other keys in a map, the tag
+-- will typically only change or be stripped away in a base case where the 'Bound' is known to be
+-- in a trivial singleton subtree with only that one key. Therefore, 'boundKey', 'minToMax', and
+-- 'maxToMin', which strip or change the tag, should be rare.
+newtype Bound t = Bound { boundKey :: Key } deriving (Eq, Ord, Show)
+
+-- | The opposite direction of a tag. This is necessary as the 'Bound's in 'IntMap_' and 'Node'
+-- are opposite each other, since they form the initial min/max bracket on the map as a whole.
+-- This is used in 'Node', where a left branch inherits the minimum from its parent, so needs
+-- to specify a new maximum (a right bound), and symmetrically in a right branch.
+type family Flipped t
+type instance Flipped L = R
+type instance Flipped R = L
+#else
+-- | A 'Key' stored somewhere within an 'IntMap'. Since this was compiled without type families,
+-- the associated tag can't be tracked properly and is therefore thrown away.
+type Bound t = Bound_
+newtype Bound_ = Bound { boundKey :: Key } deriving (Eq, Ord, Show)
+-- This, like L and R, is uninhabited to ensure that it is only used as a type parameter
+data Flipped t
 #endif
 
+data BoundOrdering = InBound | OutOfBound | Matched deriving (Eq)
 
--- A "Nat" is a natural machine word (an unsigned Int)
-type Nat = Word
+{-# INLINE compareMinBound #-}
+compareMinBound :: Key -> Bound L -> BoundOrdering
+compareMinBound k (Bound min)
+    | k > min = InBound
+    | k < min = OutOfBound
+    | otherwise = Matched
```

Comment:
As you said, I ordered the cases based on likelihood. I was very wary of using `compare` since it isn't clear what comparisons it does and they don't clearly match the cases: if I remember right, `compare` uses `<=` first. Also I wanted the comparisons to be symmetric between minima and maximum.

Does it actually matter? I'm not sure. Now that there is a `BoundOrdering` type, there is another variable to play with, namely the order of constructors (and intuitively, I'd expect the best situation to be the order of constructors matching the order of comparison). I vaguely remember doing some benchmarks at some point and seeing a measurable change favoring this, but I don't think I was thorough. Now that everything is centralized with `BoundOrdering`, it should be much easier to test, although still difficult since I expect the effects to be small and highly influenced by random CPU context.